controlplane/kubeadm/controllers/controller.go

```diff
@@ -318,7 +328,7 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
 		logger.Info("Scaling up control plane", "Desired", desiredReplicas, "Existing", numMachines)
 		return r.scaleUpControlPlane(ctx, cluster, kcp, controlPlane)
 	// We are scaling down
-	case numMachines > desiredReplicas:
+	case numMachines > desiredReplicas || controlPlane.UnhealthyMachines().Len() > 0:
```

Comment:
It feels a bit odd that we could potentially upgrade or scale up in the case that we have one or more machines in need of remediation. Should remediation be special cased to avoid those scenarios?

Reply:
I think to be more specific you are talking about a scenario like this:
- User changes replicas from 3 to 5
- One of existing machines fails health check
- According to current PR we would scale up to 5 (with one unhealthy), then scale down to 4 (remediate), then scale up to 5 again

For upgrade I think it would happen like this:
- User initiates rollout / upgrade
- We have 3 machines, one becomes unhealthy
- We scale down the unhealthy machine first due to `selectMachineForScaleDown` logic, effectively remediating it and upgrading it at the same time

Reply:
> I think to be more specific you are talking about a scenario like this:
> 
> * User changes replicas from 3 to 5
> * One of existing machines fails health check
> * According to current PR we would scale up to 5 (with one unhealthy), then scale down to 4 (remediate), then scale up to 5 again

I think the problem comes about if you consider the following scenario:
- User changes replicas from 3 to 5
- One of the existing machines fails health check in a way that causes etcd to be in a degraded state with only 2/3 replicas available
- The fourth replica is added:
  - once the bootstrapping process gets to the point that the new etcd member is added, which causes a brief loss of quorum due the new member not yet being synced, which is a scenario that we want to attempt to prevent with the exception of scaling from 1 to 2 replicas
- Prior to the fifth replica being added, another machine fails health check, also in a way that causes etcd to be unavailable locally, resulting in loss of quorum of the etcd cluster

Would we still continue to attempt to add additional replicas at this point? Even though we are at a point were no progress can be made?

Similar to above, if the fourth replica bootstraps to the point that it is added as a member to the etcd cluster, but fails to sync and join the cluster as a full member, I believe we would also be in the same situation.


> 
> For upgrade I think it would happen like this:
> 
> * User initiates rollout / upgrade
> * We have 3 machines, one becomes unhealthy
> * We scale down the unhealthy machine first due to `selectMachineForScaleDown` logic, effectively remediating it and upgrading it at the same time

Wouldn't this still attempt to scale up first, which could lead to similar scenarios as above? 

What happens if we have more than one unhealthy machine (in a way that affects etcd health), should we really be attempting to upgrade the cluster if there are any unhealthy machines?

I don't necessarily have the answers to these questions, but it seems like (at least with the information that we have on hand during reconciliation), we may have to make a choice of tradeoffs between:
- resiliency/availability of etcd (number of replicas available)
- performing the upgrade as requested
- ensuring stability when remediating (replacing remediated machine with same configuration as machine being remediated)

We could potentially chose prioritize remediation of machines over any other operation by cloning the actual config for the machine to be remediated rather than cloning templates as we do for normal scale and upgrade operations. However, we could likely only do so after the existing Machine to be remediated has been fully deleted, since there could be data like IP addresses or such in there...  We would also have to potentially strip out any information that is provider added to the spec (such as providerID).

As an aside, I'm not sure the current method of relying on the scale of the kcp to be >= 3 for remediation to happen is sufficient, we might need to make sure that machines to be remediated are not >= (actual scale/2 + 1), or better yet determine the replica count from etcd directly and compare the machines to be remediated against the member list, and:
- allow remediation of machines that are not etcd members
- allow remediation of machines that are etcd members as long as the count of those machines is not >= floor(actual scale/2 +1)